File: leave-service/src/main/java/com/otbs/leave/model/LeaveBalance.java
package com.otbs.leave.model;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDate;

@Entity
@Table(name = "leave_balance")
@Getter
@Setter
@NoArgsConstructor
public class LeaveBalance {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String userDn;

    @Column(nullable = false)
    private Integer totalLeave;

    @Column(nullable = false)
    private Integer usedLeave;

    @Column(nullable = false)
    private Integer remainingLeave;

    @Column(nullable = false)
    private LocalDate lastUpdatedDate;

    public LeaveBalance(String userDn, Integer totalLeave, Integer usedLeave, Integer remainingLeave, LocalDate lastUpdatedDate) {
        this.userDn = userDn;
        this.totalLeave = totalLeave;
        this.usedLeave = usedLeave;
        this.remainingLeave = remainingLeave;
        this.lastUpdatedDate = lastUpdatedDate;
    }

    public void addMonthlyLeave() {
        if (lastUpdatedDate != null && lastUpdatedDate.plusMonths(1).isBefore(LocalDate.now())) {
            totalLeave += 2;
            remainingLeave = totalLeave - usedLeave;
            lastUpdatedDate = LocalDate.now();
        }
    }
}

File: app/src/main/java/com/example/onetechbs/db/LeaveBalance.kt
package com.example.onetechbs.db

import com.google.gson.annotations.SerializedName
import java.time.LocalDate

data class LeaveBalance(
    @SerializedName("id")
    val id: Long? = null,

    @SerializedName("userDn")
    val userDn: String,

    @SerializedName("totalLeave")
    val totalLeave: Int,

    @SerializedName("usedLeave")
    val usedLeave: Int,

    @SerializedName("remainingLeave")
    val remainingLeave: Int,

    @SerializedName("lastUpdatedDate")
    val lastUpdatedDate: LocalDate
) {
    companion object {
        fun createEmpty(userDn: String): LeaveBalance {
            return LeaveBalance(
                userDn = userDn,
                totalLeave = 0,
                usedLeave = 0,
                remainingLeave = 0,
                lastUpdatedDate = LocalDate.now()
            )
        }
    }

    fun addMonthlyLeave(): LeaveBalance {
        val now = LocalDate.now()
        return if (lastUpdatedDate.plusMonths(1).isBefore(now)) {
            copy(
                totalLeave = totalLeave + 2,
                remainingLeave = totalLeave + 2 - usedLeave,
                lastUpdatedDate = now
            )
        } else {
            this
        }
    }

    fun calculateRemainingLeave(): Int {
        return totalLeave - usedLeave
    }

    fun hasAvailableLeave(days: Int): Boolean {
        return remainingLeave >= days
    }

    fun useLeave(days: Int): LeaveBalance {
        if (!hasAvailableLeave(days)) {
            throw IllegalStateException("Not enough leave balance")
        }
        return copy(
            usedLeave = usedLeave + days,
            remainingLeave = remainingLeave - days
        )
    }
} 